qt: Fix height of QR-less ReceiveRequestDialog

hebasto opened the pull request:
If master (89a1f7a250ef70ff2d65701564f1e24bb9280d90) is compiled without QR support, the "Request payment to..." dialog looks ugly:

![Screenshot from 2019-11-25 19-58-59](https://user-images.githubusercontent.com/32963518/69566647-3d9c1c80-0fc0-11ea-8ff6-183cea9372c5.png)

With this PR:

![Screenshot from 2019-12-26 00-42-46](https://user-images.githubusercontent.com/32963518/71451226-221c6100-277a-11ea-94ae-c19a5c6256f7.png)

![Screenshot from 2019-12-26 00-44-34](https://user-images.githubusercontent.com/32963518/71451228-29436f00-277a-11ea-8ac5-1bafe6d73b5c.png)

![Screenshot from 2019-12-26 00-48-51](https://user-images.githubusercontent.com/32963518/71451230-2e082300-277a-11ea-8c4c-726ca7b776e7.png)

Other minor changes:
- "URI" abbreviation is not translated now
- wallet name is shown if available

laanwj commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -50,7 +50,7 @@
       <size>
        <width>0</width>
-       <height>50</height>
+       <height>176</height>
```
We should really try to avoid relying on hardcoded heights. The necessary hight varies based on things such as font size, DPI, styling.

jonasschnelli replied:
Indeed. Can we make the whole dialog floating (dynamic height)?

hebasto commented:
> We should really try to avoid relying on hardcoded heights. The necessary heght varies based on things such as font size, DPI, styling

Right.

laanwj commented:
Wasn't there some other way to pack the window?

hebasto commented:
@laanwj 
> Wasn't there some other way to pack the window?

Qt allows many ways to do it. I've chosen with the minimal diff, though.

hebasto commented:
@laanwj 
> We should really try to avoid relying on hardcoded heights.

We use hardcoded sizes as defaults/minimums/maximums for all windows now.

> The necessary heght varies based on things such as font size, DPI, styling.

Currently, we rely on window resizing.

This PR changes default/minimum heights and the dialog always could be resized automagically, if QR image is supplied, or manually.

laanwj commented:
The problem is that these values can be tweaked until eternity, no two people with different OS / configuration are going to arrive at exactly the same value.

emilengler commented:
Haven't tested it yet but I think just a `this->adjustSize()` would do the job better

hebasto commented:
@emilengler 

> Haven't tested it yet but I think just a `this->adjustSize()` would do the job better

IMO, [QWidget::adjustSize()](https://doc.qt.io/qt-5/qwidget.html#adjustSize) should not be used to fix a broken layout.

It first was introduced in our repo in #16971, but I agree with @promag's https://github.com/bitcoin/bitcoin/pull/16971#issuecomment-535672268 and @Sjors's https://github.com/bitcoin/bitcoin/pull/16971#issuecomment-536469659.

hebasto commented:
@laanwj @jonasschnelli @emilengler 
Thank you for your reviews.

> Wasn't there some other way to pack the window?

Yes, alternative `QGridLayout` layout is used in the latest push.

> We should really try to avoid relying on hardcoded heights. The necessary heght varies based on things such as font size, DPI, styling.

All number in `forms/receiverequestdialog.ui` are just Qt Designer artefacts.

hebasto commented:
It seems Travis is happy now.

hebasto commented:
@practicalswift w.r.t. https://github.com/bitcoin/bitcoin/pull/17658#issuecomment-561564218 would mind reviewing this PR?

Sjors commented:
I would keep the buttons at the bottom of the dialog, consistent with other dialogs. It doesn't look very nice on macOS, especially with long translations:

<img width="703" alt="Schermafbeelding 2019-12-10 om 13 39 16" src="https://user-images.githubusercontent.com/10217/70530511-5cb7b400-1b53-11ea-8f51-3d671b555917.png">

Otherwise code in 23138de looks good and it behaves. I also tried `--without-qrencode`

hebasto commented:
@Sjors 
Thank you for your review.

> I would keep the buttons at the bottom of the dialog, consistent with other dialogs.

Fixed; also it made diffs smaller ;)

Screenshots in OP have been updated.

Sjors commented:
Much better. ACK cc88011

luke-jr commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -6,68 +6,233 @@
     </widget>
    </item>
-   <item>
-    <widget class="QTextEdit" name="outUri">
```
The QTextEdit widget made it more obvious to users they could select...

hebasto replied:
Since [`QTextEdit`](https://doc.qt.io/qt-5/qtextedit.html) inherits [`QAbstractScrollArea`](https://doc.qt.io/qt-5/qabstractscrollarea.html) with very loose size policy, it is the point of this PR to get rid of it.

luke-jr replied:
... then Concept NACK I guess

Without QTextEdit, you won't be able to select and copy everything together.

hebasto replied:
> Without QTextEdit, you won't be able to select and copy everything together.

Maybe a button "Copy ALL" could help?

luke-jr replied:
Sounds like a hack

Sjors replied:
There's a copy-URI button which copies the URI. You can tripple click on the URI itself to select it.

The `QTextEdit` used to cover all fields. Tripple click in that edit field wouldn't select the URI, but everything, which doesn't seem useful.

luke-jr commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -6,68 +6,233 @@
       <bool>true</bool>
      </property>
      <property name="textInteractionFlags">
-      <set>Qt::TextSelectableByKeyboard|Qt::TextSelectableByMouse</set>
```
Selection by keyboard can be nice...

hebasto replied:
This is not trivial, unfortunately (see #14810).

luke-jr commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -115,7 +280,7 @@
      <item>
       <widget class="QDialogButtonBox" name="buttonBox">
        <property name="standardButtons">
-        <set>QDialogButtonBox::Close</set>
```
`Close` seems more correct.

hebasto replied:
From  [`QDialogButtonBox` docs](https://doc.qt.io/qt-5/qdialogbuttonbox.html#StandardButton-enum):
- An "OK" button defined with the `AcceptRole`.
- A "Close" button defined with the `RejectRole`.

It seems the `AcceptRole` is more appropriate for interaction with this dialog.
Also see @promag's [comment](https://github.com/bitcoin/bitcoin/pull/17600#issuecomment-558392202).

luke-jr commented on src/qt/receiverequestdialog.cpp:
```diff
@@ -51,42 +45,57 @@ void ReceiveRequestDialog::setModel(WalletModel *_model)
-        html += "<b>"+tr("Message")+"</b>: " + GUIUtil::HtmlEscape(info.message) + "<br>";
-    if(model->isMultiwallet()) {
-        html += "<b>"+tr("Wallet")+"</b>: " + GUIUtil::HtmlEscape(model->getWalletName()) + "<br>";
+    ui->uri_content->setText("<a href=\"" + uri + "\">" + GUIUtil::HtmlEscape(uri) + "</a>");
```
In a label, this becomes clickable.

I'm not sure why we linkify it (so it looks like a link?), but previously it wasn't a clickable link

hebasto replied:
> I'm not sure why we linkify it (so it looks like a link?)...

It seems so.

> ... but previously it wasn't a clickable link

Agree, this change should be reverted.

hebasto replied:
> Agree, this change should be reverted.

Done in the latest push.

luke-jr commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -6,68 +6,233 @@
+      </font>
+     </property>
+     <property name="text">
+      <string notr="true">URI:</string>
```
This SHOULD be translated...

hebasto replied:
"URI" is a ubiquitous international abbreviation which does not require any translation, IMO.
Refs:
- https://www.wikipedia.org/ (in different languages)
- [`src/qt/locale/*.ts`](https://github.com/bitcoin/bitcoin/tree/master/src/qt/locale)

luke-jr replied:
What about languages which don't use Latin characters?

eg https://ur.wikipedia.org/wiki/%DB%8C%D9%88_%D8%A2%D8%B1_%D8%A2%D8%A6%DB%8C

hebasto replied:
> What about languages which don't use Latin characters?

[qt/locale/bitcoin_ur.ts](https://github.com/bitcoin/bitcoin/blob/master/src/qt/locale/bitcoin_ur.ts) does not translate "URI".

My native language does not use Latin characters, and "URI", "HTTP" are in use.

Sjors replied:
Probably best to let translators decide that. I doubt 99.9% of native English speakers even know what URI means.

luke-jr reviewed (changes requested):
Can you split the bugfix (height) from the controversial/problematic changes?

hebasto commented:
Implemented @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/17597#discussion_r362899591):
> In a label, this becomes clickable.
> I'm not sure why we linkify it (so it looks like a link?), but previously it wasn't a clickable link

Fixed tab order.

Sjors reviewed (approved):
re-ACK b8e90f3bb36706fb08aadcc2731ee9596c3d7aed; only changes in the interface file, which I don't have strong feelings about.

promag commented:
It works, but the following also works

```diff
--- a/src/qt/forms/receiverequestdialog.ui
+++ b/src/qt/forms/receiverequestdialog.ui
@@ -6,8 +6,6 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>487</width>
-    <height>597</height>
    </rect>
   </property>
   <layout class="QVBoxLayout" name="verticalLayout_3">
diff --git a/src/qt/receiverequestdialog.cpp b/src/qt/receiverequestdialog.cpp
index b4fae7d78d..cfe0bc7d52 100644
--- a/src/qt/receiverequestdialog.cpp
+++ b/src/qt/receiverequestdialog.cpp
@@ -85,6 +85,7 @@ void ReceiveRequestDialog::update()
     if (ui->lblQRCode->setQR(uri, info.address)) {
         ui->btnSaveAs->setEnabled(true);
     }
+    this->adjustSize();
 }

 void ReceiveRequestDialog::on_btnCopyURI_clicked()
```
which would make @luke-jr happy.

I think this is fine because the text area is scrollable - I mean that it's not really the same issue as the intro dialog.

hebasto commented:
@promag 
Your suggestion was the first I've tried while working on this PR.

The result look ugly for me:
![DeepinScreenshot_bitcoin-qt_20200106182801](https://user-images.githubusercontent.com/32963518/71831901-80b6d880-30b2-11ea-8f49-f2836096fb06.png)

Repo maintainers, feel free to close this PR as a controversial.

promag commented:
Please don't close, I don't think this is controversial - I prefer structured GUI instead of a big text area.

I was just replying in regards to https://github.com/bitcoin/bitcoin/pull/17597#discussion_r362965922. However I wonder if there's really the need to copy the whole text - if so I think a "copy details" button is just fine.

Only thing I can point out is if read only line edits are preferable because it's obvious the text can be copied.

hebasto commented:
@promag 
> Only thing I can point out is if read only line edits are preferable because it's obvious the text can be copied.

`QLineEdit` cannot wrap lines, by design.
Button names, "Copy ...", are self descriptive, no?

luke-jr commented:
QFontMetrics should make it practical to set a minimum width for the textedit widget?

promag commented:
> `QLineEdit` cannot wrap lines, by design.

`QLineEdit`, `QTextEdit`.. I mean one that has it's clear the user can select/tab focus and copy. But feel free to ignore the suggestion.

hebasto commented:
@luke-jr 
> QFontMetrics should make it practical to set a minimum width for the textedit widget?

It seems not convenient/accurate when different fonts/styles are mixed in the text: "**Address:** bc1..."

hebasto committed:
qt: Overhaul ReceiveRequestDialog

hebasto committed:
qt: Rename slot to updateDisplayUnit()

This commit does not change behavior.

hebasto commented:
Rebased b8e90f3bb36706fb08aadcc2731ee9596c3d7aed -> 73529f0859c060087dd41e9e176fd4198d0f385f ([pr17597.06](https://github.com/hebasto/bitcoin/commits/pr17597.06) -> [pr17597.07](https://github.com/hebasto/bitcoin/commits/pr17597.07)) due to the conflict with #15768.

jonasschnelli commented:
Tested ACK 73529f0859c060087dd41e9e176fd4198d0f385f

master:
<img width="599" alt="Bildschirmfoto 2020-05-29 um 09 53 21" src="https://user-images.githubusercontent.com/178464/83236148-582f9800-a193-11ea-912a-39e285dbaf70.png">

With this PR:
<img width="627" alt="Bildschirmfoto 2020-05-29 um 09 57 03" src="https://user-images.githubusercontent.com/178464/83236169-5fef3c80-a193-11ea-82a5-73d043473d3f.png">

Disabled QRCode support:
<img width="530" alt="Bildschirmfoto 2020-05-29 um 10 00 27" src="https://user-images.githubusercontent.com/178464/83236175-62ea2d00-a193-11ea-93de-4a7e88eaeb8e.png">
